### PR TITLE
Bug 1814187 - Refractor code to toggle the save button to active when fields are properly set.

### PR DESCRIPTION
**Pull request by gitstart**
The PR fix the bug where the save icon on `Edit login` is always active 

### Issue Screenshots
[fnx-48-issue.webm](https://user-images.githubusercontent.com/93866435/215518790-c37a2e0a-33a0-477e-8c58-42e852393f00.webm)

### Fix Screenshots
[fnx-48-fix.webm](https://user-images.githubusercontent.com/93866435/215518511-8a2b1db2-eabf-4dcc-b4ba-f3d6c8fcedc2.webm)

Pull Request checklist
### Pull Request checklist
- [x] **Quality:** This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests:** This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog:** This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/main/docs/changelog.md) or does not need one
- [x] **Accessibility:** The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/main/android/accessibility_guide.md) or does not include any user facing features

### After merge
 - Milestone: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
 - Breaking Changes: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

**GitHub Automation**
https://bugzilla.mozilla.org/show_bug.cgi?id=1814187

https://bugzilla.mozilla.org/show_bug.cgi?id=1814187
https://bugzilla.mozilla.org/show_bug.cgi?id=1814187
https://bugzilla.mozilla.org/show_bug.cgi?id=1814187
https://bugzilla.mozilla.org/show_bug.cgi?id=1814187
https://bugzilla.mozilla.org/show_bug.cgi?id=1814187
https://bugzilla.mozilla.org/show_bug.cgi?id=1814187
https://bugzilla.mozilla.org/show_bug.cgi?id=1814187
https://bugzilla.mozilla.org/show_bug.cgi?id=1814187
https://bugzilla.mozilla.org/show_bug.cgi?id=1814187
https://bugzilla.mozilla.org/show_bug.cgi?id=1814187

### CODE REVIEW

File: fenix/app/src/main/java/org/mozilla/fenix/settings/logins/fragment/EditLoginFragment.kt [hunk 1]
```diff
@@ -285,7 +304,7 @@ class EditLoginFragment : Fragment(R.layout.fragment_edit_login), MenuProvider {
         inflater.inflate(R.menu.login_save, menu)
     }
 
-    override fun onPrepareOptionsMenu(menu: Menu) {
+    override fun onPrepareMenu(menu: Menu) {
```
Comment (Mugurell) [inline fenix/app/src/main/java/org/mozilla/fenix/settings/logins/fragment/EditLoginFragment.kt:307]: nit: Changes not related to the issue should be avoid or at most be added to a separate commit.

File: fenix/app/src/main/java/org/mozilla/fenix/settings/logins/fragment/EditLoginFragment.kt [hunk 2]
```diff
@@ -154,6 +155,25 @@ class EditLoginFragment : Fragment(R.layout.fragment_edit_login), MenuProvider {
...
+                                binding.inputLayoutUsername,
+                            )
+                        }
+
+                        else -> {
+                            validUsername = true
+                            binding.inputLayoutUsername.error = null
+                            binding.inputLayoutUsername.errorIconDrawable = null
+                            binding.inputLayoutUsername.isVisible = true
+                            binding.inputLayoutUsername.isVisible = true
```
Comment (Mugurell) [inline fenix/app/src/main/java/org/mozilla/fenix/settings/logins/fragment/EditLoginFragment.kt:173]: Duplicated line.

File: fenix/app/src/main/java/org/mozilla/fenix/settings/logins/fragment/EditLoginFragment.kt [hunk 3]
```diff
@@ -264,10 +284,9 @@ class EditLoginFragment : Fragment(R.layout.fragment_edit_login), MenuProvider {
         setSaveButtonState()
     }
 
-    private fun setPasswordError() {
-        binding.inputLayoutPassword.let { layout ->
-            validPassword = false
-            layout.error = context?.getString(R.string.saved_login_password_required)
+    private fun setLayoutError(error: String?, inputLayout: TextInputLayout) {
```
Comment (Mugurell) [inline fenix/app/src/main/java/org/mozilla/fenix/settings/logins/fragment/EditLoginFragment.kt:287]: Thumbs up on making this reusable!

**Review by Mugurell — COMMENTED** [review]
Thank you for working on this!
Some small nits but this is looking and working nicely!

**Comment by gitstart** [comment]
Thank you @Mugurell for reviewing🙏.
Will soon push the requested changes. 🚀

**Comment by gitstart** [comment]
Hello @Mugurell,
Requested changes are pushed. 🚀

**Comment by gitstart** [comment]
the change is indeed related to the issue as the check mark doesn't update when `onPrepareOptionsMenu` is used.

File: fenix/app/src/main/java/org/mozilla/fenix/settings/logins/fragment/EditLoginFragment.kt (hunk 1 above)
Comment (gitstart) [reply fenix/app/src/main/java/org/mozilla/fenix/settings/logins/fragment/EditLoginFragment.kt:306]: Tthe change is indeed related to the issue as the check mark doesn't update when `onPrepareOptionsMenu` is used.

**Review by Mugurell — APPROVED** [review]
LGTM!
Thank you!

**Comment by Mugurell** [comment]
bors try

**Comment by Mugurell** [comment]
`complete-push` failed because of an intermittent UI test already tracked in https://bugzilla.mozilla.org/show_bug.cgi?id=1814579.

Trying again.

**Comment by Mugurell** [comment]
bors try

**Comment by gitstart** [comment]
> ## try
> Build failed:
> 
> * [complete-push](https://github.com/mozilla-mobile/firefox-android/runs/11550568964)

Hello @Mugurell,
Rebased with main. All Checks passed.✔
Thank you.🎉

**Comment by Mugurell** [comment]
bors try

**Comment by Mugurell** [comment]
[`CustomTabWindowFeatureTest - given a request to open window, when the url can't be handled, then handleError should be called`](https://firefoxci.taskcluster-artifacts.net/dSWqU-nQSySKLsAp5sVQSQ/0/public/reports/tests/testReleaseUnitTest/classes/mozilla.components.feature.customtabs.CustomTabWindowFeatureTest.html#given%20a%20request%20to%20open%20window,%20when%20the%20url%20can't%20be%20handled,%20then%20handleError%20should%20be%20called) failed because of 
> java.lang.IllegalStateException: Dispatchers.Main is used concurrently with setting it

which is a known intermittent issue.

---
[
`verifyLoginWithoutPasswordCanNotBeSavedTest` failed because of a seemingly unrelated issue.](https://console.firebase.google.com/u/1/project/moz-fenix/testlab/histories/bh.66b7091e15d53d45/matrices/6116009285660166965/executions/bs.520492b2971afa59/testcases/1/errors)
>  androidx.test.espresso.PerformException: Error performing 'single click' on view 'view.getId() is <2131297394/org.mozilla.fenix.debug:id/revealPasswordButton>'. 

Trying again.

**Comment by Mugurell** [comment]
bors retry

**Comment by Mugurell** [comment]
Same `verifyLoginWithoutPasswordCanNotBeSavedTest` failed again.
Posting here the stacktrace since this may be a real issue that needs more investigation:
```
androidx.test.espresso.PerformException: Error performing 'single click' on view 'view.getId() is <2131297394/org.mozilla.fenix.debug:id/revealPasswordButton>'.
	at androidx.test.espresso.PerformException$Builder.build(PerformException.java:1)
	at androidx.test.espresso.base.PerformExceptionHandler.handleSafely(PerformExceptionHandler.java:8)
	at androidx.test.espresso.base.PerformExceptionHandler.handleSafely(PerformExceptionHandler.java:9)
	at androidx.test.espresso.base.DefaultFailureHandler$TypedFailureHandler.handle(DefaultFailureHandler.java:4)
	at androidx.test.espresso.base.DefaultFailureHandler.handle(DefaultFailureHandler.java:5)
	at androidx.test.espresso.ViewInteraction.waitForAndHandleInteractionResults(ViewInteraction.java:8)
	at androidx.test.espresso.ViewInteraction.desugaredPerform(ViewInteraction.java:11)
	at androidx.test.espresso.ViewInteraction.perform(ViewInteraction.java:8)
	at org.mozilla.fenix.helpers.ViewInteractionKt.click(ViewInteraction.kt:18)
	at org.mozilla.fenix.ui.robots.SettingsSubMenuLoginsAndPasswordsSavedLoginsRobot.revealPassword(SettingsSubMenuLoginsAndPasswordsSavedLoginsRobot.kt:148)
	at org.mozilla.fenix.ui.LoginsTest$verifyLoginWithoutPasswordCanNotBeSavedTest$6.invoke(LoginsTest.kt:279)
	at org.mozilla.fenix.ui.LoginsTest$verifyLoginWithoutPasswordCanNotBeSavedTest$6.invoke(LoginsTest.kt:271)
	at org.mozilla.fenix.ui.robots.SettingsSubMenuLoginsAndPasswordRobot$Transition.openSavedLogins(SettingsSubMenuLoginsAndPasswordRobot.kt:86)
	at org.mozilla.fenix.ui.LoginsTest.verifyLoginWithoutPasswordCanNotBeSavedTest(LoginsTest.kt:271)
	... 34 trimmed
Caused by: java.lang.RuntimeException: Action will not be performed because the target view does not match one or more of the following constraints:
(view has effective visibility <VISIBLE> and view.getGlobalVisibleRect() covers at least <90> percent of the view's area)
Target view: "AppCompatImageButton{id=2131297394, res-name=revealPasswordButton, desc=Hide password, visibility=GONE, width=126, height=79, has-focus=false, has-focusable=false, has-window-focus=true, is-clickable=true, is-enabled=true, is-focused=false, is-focusable=true, is-layout-requested=true, is-selected=false, layout-params=androidx.constraintlayout.widget.ConstraintLayout$LayoutParams@YYYYYY, tag=null, root-is-layout-requested=false, has-input-connection=false, x=722.0, y=417.0}"
	at androidx.test.espresso.ViewInteraction.doPerform(ViewInteraction.java:20)
	at androidx.test.espresso.ViewInteraction.-$$Nest$mdoPerform(Unknown Source:0)
	at androidx.test.espresso.ViewInteraction$1.call(ViewInteraction.java:7)
	at androidx.test.espresso.ViewInteraction$1.call(ViewInteraction.java:1)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at android.os.Handler.handleCallback(Handler.java:938)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loop(Looper.java:223)
	at android.app.ActivityThread.main(ActivityThread.java:7656)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)

    
```

**Comment by Mugurell** [comment]
Rebased locally to get all recent changes and try again all tests.

**Comment by Mugurell** [comment]
bors try

**Comment by Amejia481** [comment]
Rebasing to the latested `main` changes.

**Comment by Amejia481** [comment]
bors try

**Comment by Amejia481** [comment]
There was an UI tests that was expecting the behaviour, we will need to update the test.

### verifyLoginWithoutPasswordCanNotBeSavedTest


```
androidx.test.espresso.PerformException: Error performing 'single click' on view 'view.getId() is <2131297462/org.mozilla.fenix.debug:id/revealPasswordButton>'.
	at androidx.test.espresso.PerformException$Builder.build(PerformException.java:1)
	at androidx.test.espresso.base.PerformExceptionHandler.handleSafely(PerformExceptionHandler.java:8)
	at androidx.test.espresso.base.PerformExceptionHandler.handleSafely(PerformExceptionHandler.java:9)
	at androidx.test.espresso.base.DefaultFailureHandler$TypedFailureHandler.handle(DefaultFailureHandler.java:4)
	at androidx.test.espresso.base.DefaultFailureHandler.handle(DefaultFailureHandler.java:5)
	at androidx.test.espresso.ViewInteraction.waitForAndHandleInteractionResults(ViewInteraction.java:8)
	at androidx.test.espresso.ViewInteraction.desugaredPerform(ViewInteraction.java:11)
	at androidx.test.espresso.ViewInteraction.perform(ViewInteraction.java:8)
	at org.mozilla.fenix.helpers.ViewInteractionKt.click(ViewInteraction.kt:18)
	at org.mozilla.fenix.ui.robots.SettingsSubMenuLoginsAndPasswordsSavedLoginsRobot.revealPassword(SettingsSubMenuLoginsAndPasswordsSavedLoginsRobot.kt:159)
	at org.mozilla.fenix.ui.LoginsTest$verifyLoginWithoutPasswordCanNotBeSavedTest$6.invoke(LoginsTest.kt:397)
	at org.mozilla.fenix.ui.LoginsTest$verifyLoginWithoutPasswordCanNotBeSavedTest$6.invoke(LoginsTest.kt:389)
	at org.mozilla.fenix.ui.robots.SettingsSubMenuLoginsAndPasswordRobot$Transition.openSavedLogins(SettingsSubMenuLoginsAndPasswordRobot.kt:107)
	at org.mozilla.fenix.ui.LoginsTest.verifyLoginWithoutPasswordCanNotBeSavedTest(LoginsTest.kt:389)
	... 34 trimmed
Caused by: java.lang.RuntimeException: Action will not be performed because the target view does not match one or more of the following constraints:
(view has effective visibility <VISIBLE> and view.getGlobalVisibleRect() covers at least <90> percent of the view's area)
Target view: "AppCompatImageButton{id=2131297462, res-name=revealPasswordButton, desc=Hide password, visibility=GONE, width=126, height=126, has-focus=false, has-focusable=false, has-window-focus=true, is-clickable=true, is-enabled=true, is-focused=false, is-focusable=true, is-layout-requested=true, is-selected=false, layout-params=androidx.constraintlayout.widget.ConstraintLayout$LayoutParams@YYYYYY, tag=null, root-is-layout-requested=false, has-input-connection=false, x=722.0, y=418.0}"
	at androidx.test.espresso.ViewInteraction.doPerform(ViewInteraction.java:20)
	at androidx.test.espresso.ViewInteraction.-$$Nest$mdoPerform(Unknown Source:0)
	at androidx.test.espresso.ViewInteraction$1.call(ViewInteraction.java:7)
	at androidx.test.espresso.ViewInteraction$1.call(ViewInteraction.java:1)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at android.os.Handler.handleCallback(Handler.java:938)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loop(Looper.java:223)
	at android.app.ActivityThread.main(ActivityThread.java:7656)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
```

**Comment by Amejia481** [comment]
@iorgamgabriel as you are putting a patch for the same bug on https://github.com/mozilla-mobile/firefox-android/pull/783, could you review this PR? No sure if we are addressing different issues.

**Comment by iorgamgabriel** [comment]
I put a bug with this GUI test https://bugzilla.mozilla.org/show_bug.cgi?id=1840561

**Comment by iorgamgabriel** [comment]
bors try

**Review by iorgamgabriel — APPROVED** [review]
